crates/accelerate/src/sabre_layout.rs

```diff
@@ -96,39 +89,59 @@ pub fn sabre_layout_and_routing(
                     run_in_parallel,
                 )
             })
-            .min_by_key(|result| result.1.map.values().map(|x| x.len()).sum::<usize>())
+            .min_by_key(|(_, result)| result.map.map.values().map(|x| x.len()).sum::<usize>())
             .unwrap()
-    };
-    (result.0, result.1, result.2.into_pyarray(py).into())
+    }
 }
 
 fn layout_trial(
-    num_clbits: usize,
-    dag_nodes: &mut Vec<(usize, Vec<usize>, HashSet<usize>)>,
+    dag: &SabreDAG,
     neighbor_table: &NeighborTable,
     distance_matrix: &ArrayView2<f64>,
     heuristic: &Heuristic,
     seed: u64,
     max_iterations: usize,
     num_swap_trials: usize,
     run_swap_in_parallel: bool,
-) -> ([NLayout; 2], SwapMap, Vec<usize>) {
+) -> ([NLayout; 2], SabreResult) {
     // Pick a random initial layout and fully populate ancillas in that layout too
     let num_physical_qubits = distance_matrix.shape()[0];
     let mut rng = Pcg64Mcg::seed_from_u64(seed);
     let mut physical_qubits: Vec<usize> = (0..num_physical_qubits).collect();
     physical_qubits.shuffle(&mut rng);
     let mut initial_layout = NLayout::from_logical_to_physical(physical_qubits);
-    let mut rev_dag_nodes: Vec<(usize, Vec<usize>, HashSet<usize>)> =
-        dag_nodes.iter().rev().cloned().collect();
+    let new_dag_fn = |nodes| {
+        // Because the current implementation of Sabre swap doesn't permute
+        // the layout when placing control flow ops, there's no need to
+        // recurse into blocks. We remove them here, but still map control
+        // flow node IDs to an empty block list so Sabre treats these ops
+        // as control flow nodes, but doesn't route their blocks.
+        let node_blocks_empty = dag
+            .node_blocks
+            .iter()
+            .map(|(node_index, _)| (*node_index, Vec::with_capacity(0)));
+        SabreDAG::new(
+            dag.num_qubits,
+            dag.num_clbits,
+            nodes,
+            node_blocks_empty.collect(),
+        )
+        .unwrap()
+    };
+
+    // Create forward and reverse dags (without node blocks).
+    // Once we've settled on a layout, we recursively apply it to the original
+    // DAG and its node blocks.
+    let mut dag_forward: SabreDAG = new_dag_fn(dag.nodes.clone());
+    let mut dag_reverse: SabreDAG = new_dag_fn(dag.nodes.iter().rev().cloned().collect());
```

Comment:
So by my understanding, we're not recursively recursing into the control-flow blocks when considering the routing?  We'll still come out with valid layouts for sure, but I think the quality will be affected because of that.  I think we can look at improving that in follow-ups, though.

Reply:
That's correct. The reason for this is that we _always_ route back to the layout from before the control flow op was placed (for now). So there's no point to recurse into the blocks when routing, since it doesn't affect the layout.

Reply:
There is a purpose to it - the quality of the Sabre layout algorithm is a bit compromised if we do it like this.  As an extreme case, consider an arbitrary circuit A, then compare to circuit B, which is just circuit A stuffed inside an `IfElseOp`.  When we're calling `SabreLayout` on circuit A, we'd apply the normal forwards-and-backwards routing attempts.  When laying out circuit B, we'd effectively just call the forwards pass, and just do it twice as many times; there'd be no influence from the forwards-and-backwards routing to tend the initial layout to something that's efficient.

I think this is fine for an initial implementation, but I do think it's likely to compromise quality for large control-flow blocks.

Reply:
Just to be clear: what you're saying is why the output is _valid_ if we don't recurse (which I agree with), what I'm saying is that by not recursing we're almost certainly compromising quality, but we can potentially worry about it later.

Reply:
> what I'm saying is that by not recursing we're almost certainly compromising quality, but we can potentially worry about it later.

At least with the current implementation of Sabre _swap_, I don't think we're compromising on quality due to skipping the recursive step. Here's how I'm thinking about it:

When we route the circuit forwards and encounter a control flow op, we route each of its blocks, and then for each of those blocks, we generate a swap epilogue that gets the block's layout back to what it was before we encountered the containing control flow op. The operation that immediately follows the control flow op in the root circuit therefore starts with the same layout that _it_ did.

When we go backwards and encounter a control flow op, the same thing happens, though we're of course starting with a different layout. If we route the inner blocks themselves backwards, we'll still generate a swap epilogue that gets us back to the layout we had when we encountered the containing control flow op.

Because we know this for both the forward and backward cases, we don't need to run routing on the inner blocks—we know where we will end up: where we started.

If we did _not_ unconditionally route all blocks back to the layout we had going into them, then we'd certainly need to recurse for best performance. But as it is, based on the way I've implemented Sabre swap, recursing would result in the same layout.

Your specific example is an excellent one, since it highlights the downside of not using the contents of inner blocks to influence the starting layout. But, this is not due to a lack of recursing into blocks, but rather that we always route blocks back to the layout we had going into their containing control flow op.

Reply:
Ahhh yeah, I get what you're saying.  Yeah, you're right: provided each control-flow block starts and ends at the same layout, it doesn't matter for the final layout what you do with the internal routing of it.  I'd missed that nuance, sorry.

Even with that, I think we lose some ability to determine quality between different parallel layout trials, because we don't gain information on how many swaps were added inside each block - for example, in my extreme example of the whole circuit stuck inside a conditional, I think the score of all initial layouts would come out the same (zero), and we'd not be able to distinguish a perfect layout from one that's maximally bad.  But you're definitely right that for a given initial layout, we'll get the same final layout under the current algorithm whether or not we recurse in.

I'm still totally fine to merge this as-is.

Reply:
Yeah, that's accurate. With the current implementation, each block is routed without influence from the blocks it contains (in control flow ops). For the root block, this means that layout inherently isn't influenced by nested blocks.

While not ideal in cases like the one you've described, I figured this approach would be the simplest low-risk way to enable control flow for this release without significantly changing how Sabre swap does routing. I've got an extension to this approach cooking that won't suffer from the same issue that we can incorporate in a future PR 😄.

Reply:
Yeah, I think you've made exactly the right choice - I thought that even when I was busy being wrong above, and now I definitely think that haha!